pkg/api/htpasswd.go

```diff
@@ -0,0 +1,184 @@
+		return fmt.Errorf("error occurred while reading creds-file: %w", err)
+	}
+
+	return nil
+}
+
+// Get returns the password associated with the login and a bool
+// indicating whether the login was found.
+// It does not check whether the user's password is correct.
+func (hc *HtpasswdClient) Get(login string) (string, bool) {
+	return hc.credMap.Get(login)
+}
+
+// Set sets the new password. It does not perform any checks,
+// the only error is possible is encryption error.
+func (hc *HtpasswdClient) Set(login, password string) error {
+	passphrase, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
+	if err != nil {
+		return fmt.Errorf("error occurred while cheking passwords: %w", err)
+	}
+
+	return hc.credMap.Set(login, string(passphrase))
+}
+
+// CheckPassword checks whether the user has a specified password.
+// It returns an error if the user is not found or passwords do not match,
+// and returns the nil on passwords match.
+func (hc *HtpasswdClient) CheckPassword(login, password string) error {
+	passwordHash, ok := hc.Get(login)
+	if !ok {
+		return zerr.ErrBadUser
+	}
+
+	err := bcrypt.CompareHashAndPassword([]byte(passwordHash), []byte(password))
+	if err != nil {
+		return zerr.ErrPasswordsDoNotMatch
+	}
+
+	return nil
+}
+
+// ChangePassword changes the user password.
+// It accepts user login, his supposed old password for verification and new password.
+func (hc *HtpasswdClient) ChangePassword(login, supposedOldPassword, newPassword string) error {
+	if len(newPassword) == 0 {
+		return zerr.ErrPasswordIsEmpty
+	}
+
+	hc.credMap.rw.RLock()
+	oldPassphrase, ok := hc.credMap.m[login]
+	hc.credMap.rw.RUnlock()
+
+	if !ok {
+		return zerr.ErrBadUser
+	}
+
+	// given old password must match actual old password
+	if err := bcrypt.CompareHashAndPassword([]byte(oldPassphrase), []byte(supposedOldPassword)); err != nil {
+		return zerr.ErrOldPasswordIsWrong
+	}
+
+	// if passwords match, no need to update file and map, return nil as if operation is successful
+	if err := bcrypt.CompareHashAndPassword([]byte(oldPassphrase), []byte(newPassword)); err == nil {
+		return nil
+	}
+
+	// encrypt new password
+	newPassphrase, err := bcrypt.GenerateFromPassword([]byte(newPassword), bcrypt.DefaultCost)
+	if err != nil {
+		return fmt.Errorf("error occurred while encrypting new password: %w", err)
+	}
+
+	file, err := os.ReadFile(hc.filepath)
+	if err != nil {
+		return fmt.Errorf("error occurred while reading creds-file: %w", err)
+	}
+
+	// read passwords line by line to find the corresponding login
+	lines := strings.Split(string(file), "\n")
+	for i, line := range lines {
+		if tokens := strings.Split(line, ":"); len(tokens) == htpasswdValidTokensNumber {
+			if tokens[0] == login {
+				lines[i] = tokens[0] + ":" + string(newPassphrase)
+
+				break
+			}
+		}
+	}
+
+	// write new content to file
+	output := strings.Join(lines, "\n")
+
+	err = os.WriteFile(hc.filepath, []byte(output), constants.DefaultDirPerms)
```

Comment:
I suggest that you first write to a temp file then just rename it so that the operation would be more atomic.

os.Rename is atomic